src/lib/operators/join_index.cpp

```diff
@@ -202,40 +208,85 @@ std::shared_ptr<const Table> JoinIndex::_on_execute() {
+    // Here we prefer to use table indexes if the join supports them. If no table index exists or other predicates than
+    // Equals or NotEquals are requested, chunk indexes are used. If no chunk index exists, NestedLoopJoin is used as a
+    // fallback solution.
+    const auto& table_indexes = _index_input_table->get_table_indexes(_index_column_id);
+
+    // table-based index join
+    if (!table_indexes.empty() && (_adjusted_primary_predicate.predicate_condition == PredicateCondition::Equals ||
+                                   _adjusted_primary_predicate.predicate_condition == PredicateCondition::NotEquals)) {
+      const auto chunk_count_index_input_table = _index_input_table->chunk_count();
+      // We need an order-preserving container here to check whether a chunk of the probe table is not indexed. For
+      // this, we assume (1) the probe table's chunk ids and (2) all indexed chunk ids to be sorted. Starting with
+      // the first elements of these sequences, we successively iterate through the elements and compare the current
+      // values.
+      auto total_indexed_chunk_ids = std::set<ChunkID>{};
+
+      // We do not take multiple table indexes created for the same column into account. For now, we only use the first
+      // available table index. Theoretically, multiple table index storing index entries for different chunk subsets,
+      // e.g., ChunkIDs{1,3} and ChunkIDs{0,2,7}, can be used in combination.
+      if (table_indexes.size() > 1) {
+        PerformanceWarning("There are multiple table indexes available, but only the first one is used.");
+      }
 
-      const auto& indexes =
-          index_chunk->get_indexes(std::vector<ColumnID>{_adjusted_primary_predicate.column_ids.second});
+      const auto& table_index = table_indexes.front();
+      const auto& indexed_chunk_ids = table_index->get_indexed_chunk_ids();
+      total_indexed_chunk_ids.insert(indexed_chunk_ids.begin(), indexed_chunk_ids.end());
 
-      if (!indexes.empty()) {
-        // We assume the first index to be efficient for our join
-        // as we do not want to spend time on evaluating the best index inside of this join loop
-        const auto& index = indexes.front();
+      _scan_probe_side_input([&](auto probe_iter, const auto probe_end, const auto probe_chunk_id) {
+        _data_join_two_segments_using_table_index(probe_iter, probe_end, probe_chunk_id, table_index);
+      });
 
-        // Scan all chunks from the probe side input
-        const auto chunk_count_probe_input_table = _probe_input_table->chunk_count();
-        for (ChunkID probe_chunk_id{0}; probe_chunk_id < chunk_count_probe_input_table; ++probe_chunk_id) {
-          const auto chunk = _probe_input_table->get_chunk(probe_chunk_id);
-          Assert(chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
+      index_joining_duration += timer.lap();
+      ++join_index_performance_data.chunks_scanned_with_index;
```

Comment:
The semantics here are different than intended: before, the counter indicated how many chunks of the indexed side were scanned using an index. Now, the result is only binary (0 or 1) and indicates whether a table index was used for the join operation.

This should be aligned with the original intention (see the behavior of the chunk-based index join execution).